### PR TITLE
[feature](Nereids) Support lots of aggregate functions

### PR DESCRIPTION
**Pull request by 924060929**
# Proposed changes
1. generate lots of aggregate functions
2. support `group_concat(columns order by order_columns)`  grammer
3. support and generate array aggregate/scalar functions, like `array_union`. we should support array grammar in the future, e.g. `select [1, 2, 3]`
4. add `checkLegalityBeforeTypeCoercion` and `checkLegalityAfterRewrite` function to check the legality of expression before type coercion and after rewrite, copy the semantic check of `FunctionCallExpr` to the checkLegality; remove the `ForbiddenMetricTypeArguments`; move the check of aes/sm4 crypto function from translator to checkLegalityBeforeTypeCoercion
5. refactor the `NullableAggregateFunction`: distinct is the first parameter, alwaysNullable is the second parameter; Fix some wrong initialize order: some function invoke super(distinct, alwaysNullable) but some function invoke super(alwaysNullable, distinct)

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
6. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
7. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
8. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
9. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 37.96 seconds
 load time: 516 seconds
 storage size: 17122043228 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230106040929_clickbench_pr_74565.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/AvgWeighted.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * AggregateFunction 'avg_weighted'. This class is generated by GenerateFunction.
+ */
+public class AvgWeighted extends AggregateFunction
+        implements BinaryExpression, ExplicitlyCastableSignature, PropagateNullable {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/AvgWeighted.java:43]: all PropagateNullable aggregate function should extends from NullableAggregateFunction

File: fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java [hunk 2]
```diff
@@ -57,6 +59,27 @@ public abstract class DataType implements AbstractDataType {
             .put(FloatType.class, () -> DoubleType.INSTANCE)
             .build();
 
+    private static final Map<Class<? extends DataType>, Promotion<DataType>> FULL_PRIMITIVE_TYPE_PROMOTION_MAP
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/types/DataType.java:62]: need some comment to explain the mean of ordinal in promotion list

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 3]
```diff
@@ -1551,4 +1559,21 @@ public DataType visitPrimitiveDataType(PrimitiveDataTypeContext ctx) {
         ctx.INTEGER_VALUE().stream().map(ParseTree::getText).forEach(l::add);
         return DataType.convertPrimitiveFromStrings(l);
     }
+
+    private Expression parseFunctionWithOrderKeys(String functionName, boolean isDistinct,
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:1567]: currently we only support the group_concat, we can consider generality if more function need order by

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/AvgWeighted.java (hunk 1 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/agg/AvgWeighted.java:43]: I think we should refactor NullableAggregateFunction in the next PR.

**Review by morrySnow — APPROVED** [review]

**Pull request merged**